text/3416-feature-metadata.md

```diff
@@ -0,0 +1,159 @@
+possibilities:
+
+```toml
+[features]
+# Current configuration will continue to work
+foo = []
+# New configurations
+bar = { enables = ["foo"], doc = "simple docstring here"}
+baz = { enables = ["foo"], public = false}
+qux = { enables = [], deprecated = true }
+quux = { enables = [], deprecated = { since = "1.2.3", note = "don't use this!" } }
+
+# Features can also be full tables if descriptions are longer
+[features.corge]
+enables = ["bar", "baz"]
+doc = """
+# corge
+
+This could be a longer description of this feature
+"""
+```
+
+The `enables` key is synonymous with the existing array, describing what other
+features are enabled by a given feature. For example,
+`foo = ["dep:serde", "otherfeat"]` will be identical to
+`foo = { enables = ["dep:serde", "otherfeat"] }`
+
+All other keys are described in their individual RFCs.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+`enables` will take the place of the feature dependency array that currently
+exists. Semantics will remain unchanged.
+
+This is a required key. If there are no requirements, an empty list should be
+provided (`enables = []`). This content is already in the index.
+
+# General Implementation & Usage
+
+Use cases for this information will likely develop with time, but one of the
+simplest applications is for information output with `cargo add`:
+
+```text
+crab@rust foobar % cargo add regex
+    Updating crates.io index
+      Adding regex v1.7.3 to dependencies.
+             Features:
+             + perf             Enables all performance related features
+             + perf-dfa         Enables the use of a lazy DFA for matching
+             + perf-inline      Enables the use of aggressive inlining inside
+                                match routines
+             + perf-literal     Enables the use of literal optimizations for
+                                speeding up matches
+             + std              When enabled, this will cause regex to use the
+                                standard library
+             + unicode          Enables all Unicode features
+             - deprecated (D)   Not a real feature, but it could be
+
+    Updating crates.io index
+```
+
+Features like `aho-corasick`, `memchr`, or `use_std` would likely be
+`public = false` since they aren't listed on the crate landing page.
+
+## Implementation note: sort order
+
+In general, any tool that renders features (`rustdoc`, `cargo add`) should
+attempt to present them in the following way:
+
+- Display default features first
+- Display non-default but stable features next (can be in a separate section)
+- Display deprecated features last (can be in a separate section)
+- Do not display private features unless receiving a flag saying to do so (e.g.
+  `--document-private-items` with `rustdoc`)
+- If ordering is not preserved, present the features alphabetically
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
```

Comment:
Extending a field to add additional types can be disruptive to parsers that normally otherwise handle future extension, e.g. oli-obk/cargo_metadata#240